<small><small><i>
Este notebook foi baseado nos notebooks disponíveis em [https://gitlab.erc.monash.edu.au/andrease/Python4Maths.git]
</i></small></small>

# Comparando e decidindo

Python permite comparar os valores associados a diferentes variáveis usando **operadores de comparação**. 

Expressões que contêm operadores de comparação são chamadas **expressões lógicas** e podem ser usadas em **estruturas de decisão** para decidir o que o seu código deve fazer. 

Neste roteiro, você aprenderá a utilizar estes conceitos da linguagem Python.

## Operadores de comparação

Operadores de comparação são utilizados para construir **expressões lógicas**. 

As regras básicas sobre expressões lógicas em Python são:

* Expressões lógicas são avaliadas como verdadeiras (**True**) ou falsas (**False**).
* Todos os operadores de comparação em Python têm precedência igual. 
* Assim como em expressões aritméticas, é possível usar parênteses para mudar a ordem de avaliação de uma expressão.
* Caso haja empate de precedência, a expressão é avaliada da esquerda para a direita.

Alguns dos operadores de comparação disponíveis em Python estão listados abaixo. 

| Símbolo | Significado |
|:----:|---|
| == | Igualdade |
| !=  | Diferença |
| < | Menor |
| > | Maior |
| <=  | Menor ou igual |
| >=  | Maior ou igual |

Note que é diferente usar `==` (comparação de igualdade) e `=` (associação de variável a valor).

In [ ]:
z = 2
z == 2

True

In [ ]:
z > 2

False

In [ ]:
z <= 2

True

In [ ]:
z != 2

False

### Exercícios de fixação

1 - Leia dois números inteiros e teste se o primeiro é menor do que o segundo.

In [ ]:
num = int(input("Digite o primeiro número: "))
num1 = int(input("Digite o segundo número: "))
num < num1

Digite o primeiro número: 20
Digite o segundo número: 15


False

2 - Sorteie dois números e teste se ambos são iguais.

**Dica --** O procedimento **randint**(*menor*, *maior*) sorteia inteiros no intervalo entre menor e maior. Este procedimento não é um procedimento nativo do Python, por isso precisamos importá-lo do módulo **random**, que o fornece.

In [ ]:
from random import randint
num = randint(0,100)
num1 = randint(0,100)
print(num, num1)
print("são iguais?", num == num1)

8 62
são iguais? False


### Comparando textos

Também é possível utilizar operadores de comparação para comparar textos:
* A igualdade (==) e a diferença (!=) funcionam de forma trivial, assim como no caso dos valores numéricos.
* As demais operações consideram a ordem alfabética.


In [ ]:
nome = "Stevejobsson"
nome == "Stevejobsson"

True

In [ ]:
nome != "Facebookson"

True

In [ ]:
'abc' < 'bbc'

True

In [ ]:
'bbc' <= 'bbc'

True

In [ ]:
'bbc' <= 'bac'

False

### Exercícios de fixação

3 - Leia dois nomes e teste se o primeiro é anterior ao segundo considerando a ordem alfabética.

In [ ]:
nome = input("Digite o primeiro nome: ")
nome1 = input("Digite o segundo nome: ")
nome < nome1

Digite o primeiro nome: Andre
Digite o segundo nome: Pedro


True

4 - Teste se os textos "Python" e "python" são considerados idênticos.

In [ ]:
"Python" == "python"

False

## Operadores lógicos

É possível criar expressões lógicas mais complexas utilizando **operadores lógicos**.

Os operadores lógicos mais simples do Python são os operadores E (**and**) e OU (**or**).

As regras básicas sobre expressões que contenham operadores lógicos em Python são:

* Operadores de comparação têm maior precedência que operadores lógicos.
* O operador lógico **and** tem maior precedência que o operador **or**. 
* É possível usar parênteses para mudar a ordem de avaliação de uma expressão.
* Caso reste apenas operações de mesma precedência, a expressão passa a ser avaliada da esquerda para a direita.

| Símbolo | Significado |
|:----:|---|
| **and** | **True**, desde que ambos os lados da expressão sejam verdadeiros |
| **or**  | **True**, desde que pelo menos um lado da expressão seja verdadeiro |

In [ ]:
"a" < "b" and 5 >= 5

True

In [ ]:
"a" <= "b" and 5 > 5

In [ ]:
"a" <= "b" or 5 > 5

### Exercícios de fixação

5 - Leia um gênero e uma idade. Teste se esta combinação representa alguém em situação de obrigatoriedade militar.

**Dica --** Em tempos de "paz", o serviço militar só é obrigatório para homens maiores de idade.

In [ ]:
gen = input("Digite 1 para MASCULINO ou 2 para FEMININO: ")
ida = int(input("Qual a sua idade? "))
print("obrigatoriedade militar:", gen == "1" and ida >= 18)


6 - Leia uma idade e teste se esta pessoa está obrigada a votar.

**Dica --** Estão obrigados a votar pessoas entre 18 e 70 anos de idade.

In [ ]:
ida = int(input("Digite a sua idade: "))
print("obrigatoriedade de votação:", 70 >= ida >= 18)

Digite a sua idade: 50
obrigatoriedade de votação: True


7 - Leia uma idade e teste se para esta pessoa o voto é facultativo.

**Dica --** O voto é facultativo para pessoas entre 16 e 18 anos ou acima dos 70 anos.

In [ ]:
ida = int(input("Digite a sua idade: "))
print("Voto facultativo:", 16 >= ida >= 18 or ida > 70)

Digite a sua idade: 28
Voto facultativo: False


8 - Leia três inteiros e teste se o primeiro pertence ao intervalo delimitado pelos dois últimos.

**Observação --** Por simplicidade, suponha que os dois últimos números estão ordenados de forma crescente.

In [ ]:
n = int(input("Digite um numero: "))
n1 = int(input("Digite o segundo numero: "))
n2 = int(input("Digite o terceiro numero: "))
n1 < n < n2

Digite um numero: 5
Digite o segundo numero: 2
Digite o terceiro numero: 10


True

### Estruturas de decisão

A principal estrutura utilizada em Python para definir se um trecho de um código será executado é a estrutura **if**, ilustrada a seguir:

```python
if condicao:
    expressao
    expressao
    ...```

As regras básicas para o uso da estrutura **if** em Python são:

* O trecho de código delimitado por uma estrutura **if** é chamado bloco.
* Um bloco é caracterizado pela *identação* de suas linhas. Assim, a quantidade de espaços no começo de cada linha é bastante importante no Python.
* O bloco de uma estrutura **if** só é executado se sua condição for satisfeita. A condição é avaliada como uma expressão lógica.

In [ ]:
numero = int(input())
if numero % 2 == 0:
    print("O número", numero, "é par.")

3


### Exercícios de fixação

9 - Leia um inteiro e o imprima se ele for múltiplo de 3.

In [ ]:
num = int(input("Digite um numero: "))
if num % 3 == 0:
    print("O número", num, "é múltiplo de 3")

Digite um numero: 27
O número 27 é múltiplo de 3


10 - Informe entradas para o código abaixo que façam a mensagem ser impressa.

In [ ]:
texto = input()
numero = int(input())
if texto < "hello" and numero ** 0.5 == 8:
    print("Mensagem impressa ;)")

Erik
64
Mensagem impressa ;)


### A estrutura ```if-else```

Uma expansão da estrutura **if** é a estrutura **if-else**, que define trechos de código mutuamente exclusivos.

```python
if condicao:
    # bloco se a condição for satisfeita
    expressao
    ...
else:
    # bloco se a condição não for satisfeita
    expressao
    ...
```
Ambos os blocos de uma estrutura **if-else** devem estar *identados*, de preferência pela mesma quantidade de espaços.

In [ ]:
numero = int(input())
if numero % 2 == 0:
    print("O número", numero, "é par.")
else:
    print("O número", numero, "é ímpar.")

### Exercícios de fixação

11 - Leia um inteiro e informe se ele é ou não múltiplo de 6.

In [ ]:
num = int(input("Digite um número: "))
if num % 6 == 0:
    print("O número", num, "é múltiplo de 6")
else:
    print("O número", num, "não é múltiplo de 6")

Digite um número: 15
O número 15 não é múltiplo de 6


12 - Leia dois nomes de pessoas e os imprima em ordem alfabética.

In [ ]:
n1 = input("Digite um nome: ")
n2 = input("Digite um nome: ")
if n1 < n2:
    print(n1,",", n2)
else:
    print(n2,",", n1)

Digite um nome: joao
Digite um nome: maria
joao , maria


### A estrutura ```if-elif-else```

É possível utilizar a estrutura **if-elif-else** para especificar múltiplos fluxos de execução do algoritmo, em função de uma condição principal e de condições alternativas:

```python
if condicao_principal:
    # bloco se a condição principal for satisfeita
    expressao
    ...
elif condicao_alternativa_1:
    # bloco se a condição principal não for satisfeita mas a condição alternativa 1 for
    expressao
    ...
elif condicao_alternativa_2:
    # bloco se nem a condição principal nem a condição alternativa 1 forem satisfeitas, mas a condição alternativa 2 for
    expressao
    ...
...
else:
    # bloco se nem a condição principal nem as condições alternativas forem satisfeitas
    expressao
    ...

```

As regras básicas para o uso da estrutura **if-elif-else** em Python são:

* O bloco **if** é executado se a condição principal for satisfeita.
* Um bloco **elif** é executado se sua condição for satisfeita e as condições acima dele não forem. 
* O bloco **else** é executado se nenhuma condição for satisfeita.


In [ ]:
from random import randint

numero = randint(0,100)
chute = int(input("Chute um número entre 0 e 100: "))
if chute == numero:
    print("Parabéns, você acertou!")
elif chute < numero:
    print("Seu chute foi menor que o número sorteado!")
else:
    print("Seu chute foi maior que o número sorteado!")

### Exercícios de fixação

13 - Leia um número e teste se ele é positivo, negativo ou nulo.

### Combinando múltiplas estruturas de decisão

Também é possível combinar estruturas de decisão em múltiplos níveis, o que é conhecido como **aninhar estruturas**.

As regras básicas para aninhar estruturas de decisão em Python são:
* Não há limite com relação a quantas estruturas e em quantos níveis.
* Blocos de níveis mais profundos devem apresentar maior grau de identação. 

In [ ]:
from random import randint

numero = randint(0,100)
if numero % 6 == 0:
    print("O número", numero, "é múltiplo de 6!")
else:
    if numero % 2 == 0:
        print("O número", numero, "é múltiplo de 2, mas não de 3!")
    elif numero % 3 == 0:
        print("O número", numero, "é múltiplo de 3, mas não de 2!")
    else:
        print("O número", numero, "não é multiplo de 2 nem de 3, muito menos de 6!")

## Exercícios do URI

[1037](https://www.urionlinejudge.com.br/judge/pt/problems/view/1037) - Um histograma é um tipo de gráfico estatístico que particiona os dados disponíveis em subgrupos e informa quantos elementos existem em cada subgrupo.

Escreva um código Python que possa ser usado para construir um histograma de um conjunto de números inteiros sorteados no intervalo [0,100), isto é, valores maiores ou iguais a 0 e menores que 100. 

Por simplicidade, considere que apenas um número será sorteado. Você deve identificar a qual dos intervalos a seguir este número pertence: [0,25), [25,50), [50,75), [75,100). 

|.| Entrada | Saída |
|-|:----:|---|
| *Exemplo 1* | 5 | [0,25) | 
| *Exemplo 2* | 50 | [50,75) |
| *Exemplo 3* | 74 | [50,75) |

In [ ]:
n1 = int(input("Digite um número maior ou igual a 0 e menor que 100: "))
if 0 <= n1 < 25:
    print("[0,25)") 
else:
    if 25 <= n1 < 50:
        print("[25,50)")
    elif 50 <= n1 < 75:
        print("[50, 75)")
    else:
        if 100 <= n1: 
            print("Somente valores maiores ou iguais a 0 e menores que 100")
        else:
            print("[75, 100)")


Digite um número maior ou igual a 0 e menor que 100: 75
[75, 100)


[1049](https://www.urionlinejudge.com.br/judge/pt/problems/view/1049) - Uma árvore de classificação é uma ferramenta usada em diferentes áreas da ciência para ajudar a identificar objetos por suas características.

Considerando a árvore de classificação abaixo, escreva um código Python que identifique um animal por suas características. Mais precisamente, seu código deve receber três textos como entrada e informar ao usuário de qual animal se trata.

<img src="https://github.com/leobezerra/python-zero/blob/master/imgs/UOJ_1049_b.png?raw=1">

|.| Entrada | Saída |
|-|----|---|
| *Exemplo 1* | vertebrado<br/>mamifero<br/>onivoro | homem | 
| *Exemplo 2* | vertebrado<br/>ave<br/>carnivoro | aguia |
| *Exemplo 3* | invertebrado<br/>anelideo<br/>onivoro | minhoca |

In [ ]:
carac1 = input("Invertebrado ou vertebrado? ")
if carac1 == "vertebrado":
    carac2 = input("ave ou mamífero? ")
    if carac2 == "ave":
        carac3 = input("carnívoro ou onivoro? ")
        if carac3 == "carnívoro" or carac3 == "carnivoro":
            print("aguia")
        else:
            print("pomba")    
    if carac2 == "mamífero" or carac2 == "mamifero":
        carac4 = input("onivoro ou herbivoro? ")    
        if carac4 == "onivoro":
            print("homem")
        else:
            print("vaca")
else:
    carac5 = input("inseto ou anelideo? ")
    if carac5 == "inseto":
        carac6 = input("hematofago ou herbivoro? ")
        if carac6 == "hematofago":
            print("pulga")
        else:
            print("lagarta")
    if carac5 == "anelideo":
        carac7 = input("hematofago ou onivoro? ")
        if carac7 == "hematofago":
            print("sanguessuga")
        else:
            print("minhoca")      
    

Invertebrado ou vertebrado? invertebrado
inseto ou anelideo? anelideo
hematofago ou onivoro?onivoro
minhoca


[1041](https://www.urionlinejudge.com.br/judge/pt/problems/view/1041) - Em sistemas de computação gráfica, compreender diferentes sistemas de coordenadas é fundamental.

Considere o sistema de coordenadas cartesiano apresentado na figura abaixo. Ele serve para representar pontos que apresentam duas dimensões, normalmente representados com a notação (x, y), onde x representa a coordenada horizontal e y a vertical.

<img src="https://github.com/leobezerra/python-zero/blob/master/imgs/UOJ_1041.png?raw=1">

O ponto (0,0) representa a origem dos eixos e é a intersecção entre as quatro subregiões ilustradas na figura abaixo (Q1, Q2, Q3 e Q4, onde Q significa *quadrante*, isto é, uma subregião). Cada quadrante da figura representa pontos com as seguintes características:

| Quadrante | Horizontal | Vertical |
|:-:|:-:|:-:|
| Q1 | x > 0 | y > 0 |
| Q2 | x < 0 | y > 0 |
| Q3 | x < 0 | y < 0 |
| Q4 | x > 0 | y < 0 |

Também é possível que um ponto esteja sobre um dos eixos:
* Um ponto estará sobre o eixo x se y = 0;
* Um ponto estará sobre o eixo y se x = 0.

Considerando este contexto, escreva um código Python que leia um ponto (duas coordenadas reais) e informe a primeira condição abaixo que se mostrar verdadeira:
* Se o ponto está na "Origem";
* Se o ponto está sobre o "Eixo X";
* Se o ponto está sobre o "Eixo Y";
* Em qual quadrante o ponto está.

|.| Entrada | Saída |
|-|:-:|:-:|
| *Exemplo 1* | 4.5<br/>-2.2 | Q4 | 
| *Exemplo 2* | 0.1<br/>0.0 | Eixo X |
| *Exemplo 3* | 0.0<br/>0.0 | Origem |

In [5]:
# declarar coordenadas

ponto_x = float(input('Digite as coordenadas do ponto x: '))
ponto_y = float(input('Digite as coordenadas do ponto y: '))

# informar condições
if ponto_x == 0 and ponto_y == 0:
    print('O ponto está na Origem.')
elif ponto_x != 0 and ponto_y == 0:
    print('O ponto está sobre o Eixo X.')
elif ponto_x == 0 and ponto_y != 0:
    print('O ponto está sobre o Eixo Y.')
elif ponto_x > 0 and ponto_y > 0:
    print('O ponto está no primeiro quadrante.')
elif ponto_x < 0 and ponto_y > 0:
    print('O ponto está no segundo quadrante.')
elif ponto_x < 0 and ponto_y < 0:
    print('O ponto está no terceiro quadrante.')
elif ponto_x > 0 and ponto_y < 0:
    print('O ponto está no quarto quadrante.')
else:
    print('O ponto não se enquadra em nenhuma condição específica.')


Digite as coordenadas do ponto x: 0.0
Digite as coordenadas do ponto y: 0.0
O ponto está na Origem.


[1051](https://www.urionlinejudge.com.br/judge/pt/problems/view/1051) - Sistemas contábeis e de recursos humanos precisam conhecer bem a legislação tributária do país para calcular os impostos a serem recolhidos mensalmente sobre o salário dos funcionários. 

Trabalhadores também precisam entender quanto ganham e porque existem descontos em seus contra-cheques.

Neste exercício, você vai ajudar empresas e trabalhadores a entender os descontos da previdência social (INSS).

A tabela usada para calcular esses descontos é apresentada a seguir.

| Salário de contribuição | Alíquota |
|:-:|:-:|
| Até 1.693,72 | 8% | 
| De 1.693,73 a 2.822,90 | 9% |
| De 2.822,91 até 5.645,80 | 11% |

Como você pode ver, o desconto sobre o salário depende da faixa salarial na qual um trabalhador se enquadre.

Além disso, existe um teto na tabela que determina que pessoas que ganhem a partir de 5.645,81 devem contribuir com 621,04 (11% de 5.645,80).

Escreva um código Python que receba um salário bruto (antes do desconto) e informe o valor da contribuição para o INSS e o salário líquido (após o desconto).

|.| Entrada | Saída |
|-|:-:|-|
| *Exemplo 1* | 7456.57 | 621.04 6835.53 | 
| *Exemplo 2* | 2822.90 | 254.06 2568.84 |
| *Exemplo 3* | 1693.73 | 152.44 1541.29 |

In [13]:
# informe de rendimentos
salario_bruto = float(input('Digite seu salário bruto: '))

# calculos da contribuição
if salario_bruto <= 1693.72:
    print(f'{0.08*salario_bruto:.2f}, {salario_bruto -(0.08*salario_bruto):.2f}')
elif 1693.72 < salario_bruto <= 2822.90:
    print(f'{0.09*salario_bruto:.2f}, {salario_bruto -(0.09*salario_bruto):.2f}')
elif 2822.90 < salario_bruto <= 5645.80:
    print(f'{0.11*salario_bruto:.2f}, {salario_bruto -(0.11*salario_bruto):.2f}')
else:
    print(f'{0.11*5645.81:.2f}, {salario_bruto -(0.11*5645.81):.2f}')

Digite seu salário bruto: 7456.57
621.04, 6835.53


Um segundo tipo de dedução mensal que incide sobre os ganhos dos trabalhadores é o *imposto de renda sobre pessoa física* (IRPF).

Para o cálculo do IRPF, utiliza-se como base de cálculo o salário bruto descontado da contribuição do INSS. 

As principais diferenças em relação ao cálculo do INSS são:
* Quem possui salário-base até 1.903,98 é isento do pagamento do IRPF.
* Não há teto para a contribuição do IRPF.
* O valor do IRPF cobrado a uma pessoa é abatido de um valor que varia em função da alíquota correspondente.

A tabela usada para o cálculo do IRPF é mostrada abaixo.

| Base de cálculo (R\$) | Alíquota | Parcela a deduzir |
|:-:|:-:|:-:|
| Até 1.903,98 | - | - |
| De 1.903,99 até 2.826,65 | 7,5% | 142,80 |
| De 2.826,66 até 3.751,05 | 15% | 354,80 |
| De 3.751,06 até 4.664,68 | 22,5% | 636,13 |
| Acima de 4.664,68 | 27,5% | 869,36 |

Para entender melhor o cálculo do IRPF, vamos tomar como exemplo uma pessoa que, descontada a contribuição do INSS, ganha 2.000,00.

A alíquota para esta remuneração é de 7,5% e o abatimento de 142,80. Assim, a contribuição do IRPF será de 2.000,00 * 7,5% - 142,80 = 7,20.

Escreva um código Python que receba um salário-base (já descontado do INSS) e informe o valor da contribuição do IRPF e o salário líquido (após o desconto do IRPF).

|.| Entrada | Saída |
|-|:-:|-|
| *Exemplo 1* | 6835.53 | 1010.41 5825.12 | 
| *Exemplo 2* | 2568.84 | 49.86 2518.98 |
| *Exemplo 3* | 1541.29 | 0.00 1541.29 |